### PR TITLE
Hypre ADS

### PR DESCRIPTION
**Pull request by xiaodizhang29**
Added Hypre ADS preconditioner

### CODE REVIEW

**Review by wence- — CHANGES_REQUESTED** [review]
Thanks! I just had some minor changes (listed in the comments).

File: firedrake/preconditioners/hypre_ads.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        pc = PETSc.PC().create(comm=obj.comm)
+        pc.incrementTabLevel(1, parent=obj)
+        pc.setOptionsPrefix(prefix + "hypre_ads_")
+        pc.setOperators(A, P)
+
+        pc.setType('hypre')
+        pc.setHYPREType('ads')
+        pc.setHYPREDiscreteGradient(G)
+        pc.setHYPREDiscreteCurl(C)
+        pc.setCoordinates(mesh.coordinates.dat.data_ro)
```
[the remark this answers is not in the document]
Comment (xiaodizhang29) [reply firedrake/preconditioners/hypre_ads.py:40]: Yes, you are right. I will resolve it.

File: tests/regression/test_hypre_ads.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+              'pc_type': 'python',
+              'pc_python_type': 'firedrake.HypreADS',
+              }
+
+    u = Function(V)
+    solve(a == L, u, bc, solver_parameters=params)
+    assert (errornorm(Constant((1, 0.5, 4)), u, 'L2') < 1e-10)
+
+
+@pytest.mark.skipcomplex
```
[the remark this answers is not in the document]
Comment (xiaodizhang29) [reply tests/regression/test_hypre_ads.py:33]: OK!

**Review by wence- — APPROVED** [review]
Thanks, I'll wait for Jenkins to be happy, but otherwise looks good!

File: firedrake/preconditioners/hypre_ads.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        pc.incrementTabLevel(1, parent=obj)
+        pc.setOptionsPrefix(prefix + "hypre_ads_")
+        pc.setOperators(A, P)
+
+        pc.setType('hypre')
+        pc.setHYPREType('ads')
+        pc.setHYPREDiscreteGradient(G)
+        pc.setHYPREDiscreteCurl(C)
+        V = VectorFunctionSpace(mesh, "Lagrange", 1)
+        linear_coordinates = Interpolator(SpatialCoordinate(mesh), V).dat.data_ro.copy()
```
Comment (wence-) [inline firedrake/preconditioners/hypre_ads.py:41]: Ah sorry, we have to import `SpatialCoordinate` from `ufl` at the top of the file for this to work.

Comment (xiaodizhang29) [reply firedrake/preconditioners/hypre_ads.py:41]: OK, I will resolve it right now.

Comment (xiaodizhang29) [reply firedrake/preconditioners/hypre_ads.py]: I have got lost. Could you help me?

Comment (wence-) [reply firedrake/preconditioners/hypre_ads.py]: The new changes look to be correct. And the only failing test I now see is unrelated to your changes (and is a problem we are trying solve separately).

Comment (xiaodizhang29) [reply firedrake/preconditioners/hypre_ads.py]: Many thanks.

**Comment by wence-** [comment]
Thanks for your patience. I merged the test fixes that hopefully fix the failures we observe. Can you please merge `master` into this branch (or rebase this branch on top of master).

**Comment by wence-** [comment]
Thanks very much!

**Pull request merged**